readme.txt
The Binding of Madou
Beta v0.1.1
6/3/2023

==CREDITS==
Coder: 
wither2x

Designers: 
Sadship
wither2x

Sprites:
Taoyufei
wither2x

==Introduction==
This is a fan-made mod based on Madou Monogatari and Puyo Puyo series, which features 1 new player, 2 new items and 1 new challenge.
This mod is still under development. More new content is planned to be added soon, including collectibles, trinkets, characters, bosses, and story.
The program part of this mod needs to be improved. You are welcome to make suggestions for its improvement.
This mod supports local co-op mode and controller.
This mod has compatibility with External Item Descriptions and Mod Config Menu.

==New Character==
◇Arle Nadja
  ▷Starting attributes
    Health: 3 red hearts
    Speed: 0.9
    Tears: 2.53
    Damage: 3
    Range: 5.5
    Shot Speed: 1
    Luck: 0
  ▷Starting items
    Blue grimoire (Pocket active item)
    x20 Mana points
  ▷Features
   ▷Level & exp
    Killing enemies grants exp points to the player. The player levels up when reaches a certain amount of exp points.
    The amount of exp points obtained is related to the max health point of the enemy. Champion enemies grant 1 bonus exp point on death.
    When levels up, all red hearts and mana points will be restored, and the player will gain mp cap increase and 1 random stat increase. The possibilities include:
	+0.2 Flat damage
	+0.1 Tears up
	+0.5 Range up
	+0.05 Speed up
	+0.2 Luck up
    If the player reaches Lv.15, opens the door to the Boss Rush regardless of the timer. If the player reaches Lv.25, opens the door to the Blue Womb regardless of the timer.
   ▷Critical hit
    The player has a chance to cause a critical hit when deals damage to enemies. Critical hit deals 200% player's Damage.
    The player has 4% Critical Chance at the start of a new run.
    The Critical Chance is positively correlated with player's Luck, with a minimum of 1%. The Teardrop Charm grants +3% Critical Chance to player.
   ▷Compatibilities:
    ·Blue Grimoire
    While held: Hearts, Coins, bombs and keys have 1/10 chance to be replaced into mana pickups, restores 20 mp after picked up. However, the mp won't be restored automatically after cleaning a room.
	Monsters have 5% chance to drop mana pickups on death if player's mp is less than 15%.
    The player can held 8 different spells.
    	Press L Alt or R Ctrl (or D-pad Left or D-pad Right on controller) to switch among these spells. If you are using keyboard, you can use the number keys (1~8) to select and use a spell directly.
    	Fire (Phase 1) and Ice Storm (Phase 1) will be unlocked automatically at the start of a new run. Other spells will be unlocked if the player reachs a certain level or has certain items or kills cettain bosses.
    On use, consumes several mp and grants the selected spell effect.It takes some spells few time to recharge before the player can use them again.
    The spells Arle can use are introduced in further detail below.
	·Fire (Will be unlocked at the start of a new run; reach Lv.7, Lv.16, and Lv.26 to upgrade)
	 Type: Aggressive
	 Effects:
	 Phase 1: The player has a chance to fire tears that burn enemies.
	 Phase 2: Grants player a chance to shoot fires alongside the regular tears. Grants immunity to fire.
	 Phase 3: Lights all enemies on fire on use. Grants immunity to explosions.
	 Phase 4: The player always fires buring tears.
	 Increases Luck if the player has Fire Mind.
	·Ice Storm (Will be unlocked at the start of a new run; reach Lv.9, Lv.18, and Lv.28 to upgrade)
	 Type: Aggressive
	 Effects:
	 Phase 1: The player has a chance to fire tears that slow enemies.
	 Phase 2: Grants player a chance to fire iced tears like Uranus.
	 Phase 3: Freezes all enemies on use.
	 Phase 4: The player always fires iced tears.
	 Increases Tears if the player has Uranus.
	·Thunder (Reach Lv.5, Lv.11, Lv.22, and Lv.30 to unlock or upgrade)
	 Type: Aggressive
	 Effects:
	 Phase 1: Fire static tears like Technology Zero.
	 Phase 2: Fire electric tears like Jacob's Ladder.
	 Phase 3: Grants player a chance to fire confusion tears.
	 Phase 4: Grants player a chance to fire petrifying tears.
	 Increases Damage if the player has Technology Zero or Jacob's Ladder.
	·Healing (Reach Lv.2 to unlock)
	 Type: Defensive
	 Effects:
	 Heals one red heart on use. Grants a half soul heart at full hp.
         In co-op mode, heals all other co-op players for one red heart.
	·Diacute (Reach Lv.3, Lv.13, Lv.24 or defeat Mom's Heart (or It Lives!) to unlock or upgrade)
	 Type: Special
	 Effects:
	 Spawns a light orb orbital that mimic the player's shoot direction and tears. Lasts for 30 seconds.
	 The tears will inherite all tear effects of player except special weapons like Brimstone and Mom's Knife.
	 The upper limit of the amount of orbs equivalent to the phase of this spell. The player may only have up to 4 orbs within a spell cycle.
	 Compatibilities with other spells:
	    Healing: Decrease the recharging time of Healing spell.
	    Bayoen: If the enemies have a tainted veasion, the spell has a chance to morphs them into tainted versions before charming them. It is invalid for the enemies spawned by bosses.
	    Revia: Stacks multiple Holy Mantle effects to player. Deals more damage to enemies when the Holy Mantle is broken.
	    Jugem: Grants immunity to explosions.
	·Bayoen (Reach Lv.14 to unlock)
	 Type: Defensive
	 Effects:
	 Adds a permanent charmed effect to all enemies. They will follow the player to different rooms.
	 The bosses and enemies spawned by bosses makes the effect temporary instead.
	·Revia (Reach Lv.20 to unlock)
	 Type: Defensive
	 Effects:
	 Grants a temporary Holy Mantle effect for current room. Makes player invincible for 10 seconds if the player already has Holy Mantle effect.
	 When the Holy Mantle is broken, deals 30 + 5 * player's Damage to all enemies.
	·Jugem (Obtain the Green Grimoire to unlock)
	 Type: Lock-on
	 Effects:

change_log.txt
◇2023/4/3  Alpha v0.0.0
完成了第一个内测版，仅包含一个新角色和两个新道具。

◇2023/4/8  Alpha v0.0.1
新增内容：现已支持英文EID。
角色【阿露露】修复：修复了击败妈腿时恰好升到15级的情况下未能第一时间开启Boss Rush的错误。
角色【阿露露】修复：修复了击败妈腿时获得的经验值数目异常的bug。
法术【火炎术】增强：现在达到3级即可免疫爆炸伤害。
法术【气爆弹】改动：效果由直接引爆改为召唤导弹；爆炸现在造成2000%角色伤害，但会伤到角色；解锁该法术后，如果持有【史诗胎儿博士】，则可免疫所有爆炸伤害。
杂项修复：修正了个别法术的介绍文本。

◇2023/4/11  Alpha v0.0.2
新增内容：加入了存档功能。
道具【蓝色魔导书】改动：现在使用攻击性法术时和拾取魔导力时，角色将会短暂闪烁一次。
法术【二阶强化术】改动：现在光球消失时会显示动画。
法术【二阶强化术】修复：修复了在拥有光球的情况下重开后光球不消失的bug。
法术【气爆弹】改动：现在会优先锁定离角色近的目标。

◇2023/4/14  Alpha v0.0.3
角色【阿露露】修复：修复了在满足等级条件时，凹凸层的入口未能正常生成的错误。
道具【蓝色魔导书】修复：阿露露清理房间后不再自动回复魔导力，其他角色不受影响。
道具【蓝色魔导书】修复：修复了其他角色在同一房间内多次使用法术时造成的属性加成混乱之bug。
道具【蓝色魔导书】修复：金硬币不再被转化为魔导力。
道具【蓝色魔导书】改动：现已兼容【虚空】：被虚空吞噬后将取消魔导力机制，改为通过消耗虚空本身的充能来使用法术。
道具【蓝色魔导书】改动：现在使用攻击性法术时，角色脚下会出现一个法阵。
道具【蓝色魔导书】改动：优化了按键操作：对于攻击性法术、特殊法术和瞄准-锁定型法术，现在按下数字键时可一键使用或取消使用。
道具【绿色魔导书】改动：困难模式下掉率改为1/180。
法术【火炎术】改动：现在泪弹颜色与普通泪弹略有不同。
法术【气爆弹】修复：修复了一个使未被光束瞄准的敌怪也被锁定的错误。
法术【气爆弹】增强：与【二阶强化术】搭配使用时，现在可免疫爆炸伤害。

◇2023/4/20  Beta v0.1.0
“新手魔导师勇闯地下室。”
"An adventure in basement of a developing magician."
完成了第一个公测版，现已上架创意工坊。
新增内容：加入了新挑战【唤星秘术】。
道具【蓝色魔导书】修复：修复了由于浮点数精度问题导致的魔导力刚好达到需求下限但对应法术被禁用的bug。
法术【二阶强化术】改动：与【美德之书】搭配使用时，现在改为生成一个灰色普通魂火。
法术【繁花乱象】改动：与【美德之书】搭配使用时，现在改为生成一个有几率发射标记泪弹的魂火。
法术【气爆弹】削弱：【二阶强化术】不再增加导弹的爆炸伤害。
杂项修复：修正了个别法术的介绍文本。



◇2023/6/3  Beta v0.1.1

代码全面重写。
All code has been rewritten.

新增内容：现已支持本地多人模式和手柄操作。
New content: Now the mod supports local co-op mode and controller.

新增内容：加入了控制台指令“tbom_debug”。该指令可被简写为“tdb”。
New content: Added a debug console command named "tbom_debug", which can be abbreviated as "tdb".

道具【蓝色魔导书】修复：修复了使用攻击性法术期间，角色持有的刀剑类武器不停闪烁的bug。
Fixes for Blue Grimoire: Fixed an issue that will cause the knife weapon of the player to flash continuously.

道具【蓝色魔导书】修复：修复了使用攻击性法术期间，角色脚下的法阵未能如预期正确渲染的bug。
Fixes for Blue Grimoire: Fixed an issue that the magic circle won't be rendered correctly as expected.

道具【蓝色魔导书】改动：将该道具在宝箱房道具池中的权重修改为1。
Changes on Blue Grimoire: The weight of the item in Treasure Room pool is modified to 1.

道具【蓝色魔导书】改动：将该道具以0.5的权重加入了贪婪宝箱房。
Changes on Blue Grimoire: Added the item to Greed Treasure Room pool with a weight of 0.5.

道具【蓝色魔导书】改动：魔导力图标不再遮挡EID的文字和个别角色的附加物品栏。
Changes on Blue Grimoire: The icon of mp won't block the text of EID and the additional inventory of some characters.

道具【蓝色魔导书】改动：恢复中的法术现在会在HUD上显示冷却时间。
Changes on Blue Grimoire: Now the recharging time of spells will be displayed on the HUD.

道具【蓝色魔导书】改动：优化了按键操作：现在在使用法术期间亦可直接切换法术。
Changes on Blue Grimoire: Improved the key operation: Now the player can switch among the spells directly, even when the player is using current spell.

道具【蓝色魔导书】改动：角色持有卡片/药丸时，现在会暂时隐藏法术图标。
Changes on Blue Grimoire: Now the icon of spells will be temporarily hidden while the player is holding cards or pills.

道具【蓝色魔导书】改动：个别角色在触发法术时，HUD现在会显示不同于常规的名称。
Changes on Blue Grimoire: Now the HUD will display the name of spell that is different from regular names when trigger a spell as some characters.

道具【蓝色魔导书】改动：将常规角色的初始魔导力上限修改为15。

main.lua
------v0.1.2开发版------
tbom = RegisterMod("The Binding of Madou", 1)

local Fonts = {
	["en"] = Font(),
	["zh"] = Font(),
	["number"] = Font(),
}
Fonts["en"]:Load("font/pftempestasevencondensed.fnt")
Fonts["zh"]:Load("font/cjk/lanapixel.fnt")
Fonts["number"]:Load("font/pftempestasevencondensed.fnt")

tbom.modCard = {}

tbom.modChallenge = {
	CHALLENGE_METEOR_SHOWER = Isaac.GetChallengeIdByName("Meteor Shower"),
	CHALLENGE_DESCENDING_INTO_PUYO_HELL = Isaac.GetChallengeIdByName("Descending Into Puyo Hell"),
}

tbom.modCollectibleType = {
	COLLECTIBLE_BLUE_GRIMOIRE = Isaac.GetItemIdByName("Blue Grimoire"),
	COLLECTIBLE_PUYO_HAT = Isaac.GetItemIdByName("Puyo Hat"),
	COLLECTIBLE_WORMHOLE_APPLE = Isaac.GetItemIdByName("Wormhole Apple"),
	COLLECTIBLE_HEART_SHAPED_COOKIE = Isaac.GetItemIdByName("Heart Shaped Cookie"),

	COLLECTIBLE_GREEN_GRIMOIRE = Isaac.GetItemIdByName("Green Grimoire"),
}

tbom.modCostume = {
	ARLE_HAIR = Isaac.GetCostumeIdByPath("gfx/characters/character_arle_hair.anm2"),
}

tbom.modEffectVariant = {
	EXP = Isaac.GetEntityVariantByName("EXP"),
	LASER_SIGHT = Isaac.GetEntityVariantByName("Laser Sight"),
	LOCKON_MARK = Isaac.GetEntityVariantByName("Lock-on Mark"),
	MAGIC_CIRCLE = Isaac.GetEntityVariantByName("Magic Circle"),
	CHARGE_LASER_SIGHT = Isaac.GetEntityVariantByName("Charge Laser Sight"),
	PUYO_FIRE_POINT = Isaac.GetEntityVariantByName("Puyo Fire Point"),
	PUYO_GIBS = Isaac.GetEntityVariantByName("Puyo Gibs"),
	ET_CHECKPOINT = Isaac.GetEntityVariantByName("ET Checkpoint"),
	ET_SPACESHIP = Isaac.GetEntityVariantByName("ET Spaceship"),

	BLANK_ANIM = Isaac.GetEntityVariantByName("Blank Animation"),
}

tbom.modEntityType = {
	ENTITY_PUYO = Isaac.GetEntityTypeByName("Green Puyo"),
}

tbom.modFamiliarVariant = {
	LIGHT_ORB = Isaac.GetEntityVariantByName("Light Orb"),
}

tbom.Fonts = Fonts

tbom.modPickupVariant = {
	PICKUP_MANA = Isaac.GetEntityVariantByName("Mana"),
	PICKUP_GEL = Isaac.GetEntityVariantByName("Green Gel"),
}

tbom.modPlayerType = {
	PLAYER_ORIGINAL = 0,	--仅用于相关表的检索，不代表某个特定的角色实体类型
	PLAYER_ARLENADJA = Isaac.GetPlayerTypeByName("Arle Nadja", false),
	PLAYER_ARLENADJA_B = Isaac.GetPlayerTypeByName("Tainted Arle", true),
}

tbom.modSoundEffect = {
	SOUND_EXP_GET = Isaac.GetSoundIdByName("EXP Get"),
	SOUND_LASER_SIGHT = Isaac.GetSoundIdByName("Laser Sight"),
	SOUND_LOCKON_MARK = Isaac.GetSoundIdByName("Lock-on Mark"),
	SOUND_THUNDER_BURSTING = Isaac.GetSoundIdByName("Thunder Bursting"),
	SOUND_PUYO_1_CHAIN = Isaac.GetSoundIdByName("Puyo 1 Chain"),
	SOUND_PUYO_2_CHAIN = Isaac.GetSoundIdByName("Puyo 2 Chain"),
	SOUND_PUYO_3_CHAIN = Isaac.GetSoundIdByName("Puyo 3 Chain"),
	SOUND_PUYO_4_CHAIN = Isaac.GetSoundIdByName("Puyo 4 Chain"),
	SOUND_PUYO_5_CHAIN = Isaac.GetSoundIdByName("Puyo 5 Chain"),
	SOUND_PUYO_6_CHAIN = Isaac.GetSoundIdByName("Puyo 6 Chain"),
	SOUND_PUYO_7_CHAIN = Isaac.GetSoundIdByName("Puyo 7 Chain"),
	SOUND_PUYO_BURST = Isaac.GetSoundIdByName("Puyo Burst"),
	SOUND_GEL_GET = Isaac.GetSoundIdByName("Gel Get"),
	SOUND_GEL_GET_ARCADE_1 = Isaac.GetSoundIdByName("Gel Get Arcade 1"),
	SOUND_GEL_GET_ARCADE_2 = Isaac.GetSoundIdByName("Gel Get Arcade 2"),
	SOUND_GEL_GET_ARCADE_3 = Isaac.GetSoundIdByName("Gel Get Arcade 3"),
	SOUND_GREEN_GEL_HIT = Isaac.GetSoundIdByName("Green Gel Hit"),
	SOUND_ET_SPACESHIP = Isaac.GetSoundIdByName("ET Spaceship"),
}

tbom.modTrinketType = {
	TRINKET_ANOTHER_CARTRIDE = Isaac.GetTrinketIdByName("Another Cartride"),
}

tbom.MagicType = {
	NONE = 0,			--空类型法术：无CD，使用瞬间消耗魔导力
	AGGRESSIVE = 1,		--攻击性法术：无CD，使用期间每秒消耗魔导力
	DEFENSIVE = 2,		--防御性法术：有CD，使用瞬间消耗魔导力，CD期间不可使用
	SPECIAL = 3,		--特殊法术：有CD，使用瞬间消耗魔导力，CD期间某段时间内仍可使用
	LOCKON = 4,			--瞄准-锁定型法术：无专用CD，使用后每瞄准一个目标消耗魔导力
	HELPER = 5,			--辅助性法术：有短暂CD，不消耗魔导力，但使用瞬间消耗其他数值（比如血量）
}

tbom.SpellType = {
	SPELL_INVALID = -1,					--无效法术
	SPELL_NONE = 0,						--间属性法术
	SPELL_FIRE = 1,						--火炎术
	SPELL_ICE_STORM = 2,				--冰霜风暴
	SPELL_THUNDER = 3,					--闪电术
	SPELL_HEALING = 4,					--治疗术
	SPELL_DIACUTE = 5,					--二阶强化术
	SPELL_BAYOEN = 6,					--繁花乱象
	SPELL_REVIA = 7,					--荆棘障壁
	SPELL_JUGEM =  8,					--气爆弹
	SPELL_LWARK_WOID = 9,				--吸星术
	SPELL_BRAIN_DUMBED = 10,			--蚀智术
	SPELL_ILLUSION = 11,				--幻象错觉
	SPELL_HEAVEN_RAY = 12,				--天界射线
	SPELL_SINGING_VOICE_OF_HOPE = 13,	--希望之音
	SPELL_ABYSS = 14,					--狱炎术
	SPELL_CHAOS = 15,					--混沌冰暴
	SPELL_LIGHTENING = 16,				--雷电术
	SPELL_ECLIPSE = 17,					--侵腐术
	SPELL_LABYRINTH = 18,				--幻惑强化术
	SPELL_VOID_HOLE = 19,				--虚空虹洞
	SPELL_RAGNAROK = 20,				--末世权化
	SPELL_GRAND_CROSS = 21,				--灾星阵

	SPELL_FLAME_TORNADO = 98,			--卷炎术
	SPELL_DARK_BLIZZARD = 99,			--暗夜霜暴
	SPELL_HEAVENLY_THUNDER = 100,		--裂天震电
	SPELL_VOLCANA = 101,				--爆炎奥术
	SPELL_GLACIERIA = 102,				--寒霜奥术
	SPELL_LIGHTING_BRUST = 103,			--惊雷奥术
	SPELL_AURORA = 104,					--曙光奥术
}

tbom.AwardFlag = {
	AWARD_STATS_INCREASE = (1 << 0),
	AWARD_HP_UP = (1 << 1),
	AWARD_MP_UP = (1 << 2),
	AWARD_UPGRADE_SPELL = (1 << 3),
	AWARD_RECOVERY_UP = (1 << 4),
}

tbom.PuyoVariant = {
	PUYO_GREEN = Isaac.GetEntityVariantByName("Green Puyo"),
	PUYO_PURPLE = Isaac.GetEntityVariantByName("Purple Puyo"),
	PUYO_RED = Isaac.GetEntityVariantByName("Red Puyo"),
	PUYO_YELLOW = Isaac.GetEntityVariantByName("Yellow Puyo"),
	PUYO_BLUE = Isaac.GetEntityVariantByName("Blue Puyo"),
}

tbom.GelSubType = {
	GEL_GREEN = 0,
	GEL_PURPLE = 1,
	GEL_RED = 2,
	GEL_YELLOW = 3,
	GEL_BLUE = 4,
}

tbom.tbomCallbacks = {
	--带有BASE字样的回调函数具由更高的优先级
	TBOMC_ENABLE_SPELL = "TBOMC_ENABLE_SPELL",							--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），rng（RNG对象），player（角色实体对象）, use_flags（整数）；返回值类型：是否显示动画（逻辑）
	TBOMC_DISABLE_SPELL = "TBOMC_DISABLE_SPELL",						--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），rng（RNG对象），player（角色实体对象）, use_flags（整数）；返回值类型：是否显示动画（逻辑）
	TBOMC_SPELL_UPDATE = "TBOMC_SPELL_UPDATE",							--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），player（角色实体对象）；返回值类型：无
	TBOMC_ENABLE_SPELL_BASE = "TBOMC_ENABLE_SPELL_BASE",				--回调参数：magic_type（整数）；函数参数：spell_ID（整数），rng（RNG对象），player（角色实体对象）, use_flags（整数）；返回值类型：无
	TBOMC_DISABLE_SPELL_BASE = "TBOMC_DISABLE_SPELL_BASE",				--回调参数：magic_type（整数）；函数参数：spell_ID（整数），rng（RNG对象），player（角色实体对象）, use_flags（整数）；返回值类型：无
	TBOMC_SPELL_UPDATE_BASE = "TBOMC_SPELL_UPDATE_BASE",				--回调参数：magic_type（整数）；函数参数：spell_ID（整数），player（角色实体对象）；返回值类型：无

	TBOMC_USE_SPELL = "TBOMC_USE_SPELL",								--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），rng（RNG对象），player（角色实体对象）, use_flags（整数）；返回值类型：是否显示动画（逻辑）
	TBOMC_USE_SPELL_BASE = "TBOMC_USE_SPELL_BASE",						--回调参数：magic_type（整数）；函数参数：spell_ID（整数），rng（RNG对象），player（角色实体对象）, use_flags（整数）；返回值类型：无

	TBOMC_POST_SPELL_INIT = "TBOMC_POST_SPELL_INIT",					--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），player（角色实体对象）；返回值类型：无
	TBOMC_POST_SPELL_INIT_BASE = "TBOMC_POST_SPELL_INIT_BASE",			--回调参数：magic_type（整数）；函数参数：spell_ID（整数），player（角色实体对象）；返回值类型：无

	TBOMC_POST_USE_SIM_SPELL = "TBOMC_POST_USE_SIM_SPELL",				--回调参数：sim_spell_ID（整数）；函数参数：sim_spell_ID（整数），rng（RNG对象），player（角色实体对象）, use_flags（整数）；返回值类型：无
	TBOMC_PRE_BURST_ATK_CHARGE = "TBOMC_PRE_BURST_ATK_CHARGE",			--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），reberu（整数），rng（RNG对象），player（角色实体对象）；返回值类型：无
	TBOMC_POST_BURST_ATK_CHARGE = "TBOMC_POST_BURST_ATK_CHARGE",		--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），reberu（整数），timeout（整数），rng（RNG对象），player（角色实体对象）；返回值类型：无
	TBOMC_POST_TRIGGER_ATK_EFFECT = "TBOMC_POST_TRIGGER_ATK_EFFECT",	--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），rng（RNG对象），player（角色实体对象）, use_flags（整数）；返回值类型：无
	TBOMC_POST_FIRE_LOCKON_WEAPON = "TBOMC_POST_FIRE_LOCKON_WEAPON",	--回调参数：spell_ID（整数）；函数参数：spell_ID（整数），player（角色实体对象），target（实体对象）；返回值类型：无
	TBOMC_POST_COST_SPCHARGE = "TBOMC_POST_COST_SPCHARGE",				--回调参数：无；函数参数：player（角色实体对象），attribute_name（字符串），amount（正浮点数）；返回值类型：无
	TBOMC_PRE_GET_UPGRADE_AWARD = "TBOMC_PRE_GET_UPGRADE_AWARD",		--回调参数：无；函数参数：player（角色实体对象），award_flag（整数）；返回值类型：无
	
	TBOMC_DOUBLE_TAP = "TBOMC_DOUBLE_TAP",								--回调参数：button_action（整数）；函数参数：button_action（整数），player（角色实体对象）；返回值类型：无
	TBOMC_TAP_AND_HOLD_MOVING = "TBOMC_TAP_AND_HOLD_MOVING",			--回调参数：action_type（整数；0:待机；1：按住；2：释放）；函数参数：move_dir（向量），player（角色实体对象）；返回值类型：无
	TBOMC_TAP_AND_HOLD_SHOOTING = "TBOMC_TAP_AND_HOLD_SHOOTING",		--回调参数：action_type（整数；0:待机；1：按住；2：释放）；函数参数：shoot_dir（向量），player（角色实体对象）；返回值类型：无
	TBOMC_PRE_ADD_COLLECTIBLE = "TBOMC_PRE_ADD_COLLECTIBLE",			--回调参数：collectible_type（整数）；函数参数：collectible_type（整数），rng（RNG对象），player（角色实体对象）；返回值类型：无
	TBOMC_POST_ADD_COLLECTIBLE = "TBOMC_POST_ADD_COLLECTIBLE",			--回调参数：collectible_type（整数）；函数参数：collectible_type（整数），rng（RNG对象），player（角色实体对象），is_newly_added（逻辑）；返回值类型：无
	TBOMC_PRE_ADD_TRINKET = "TBOMC_PRE_ADD_TRINKET",					--回调参数：trinket_type（整数）；函数参数：trinket_type（整数），rng（RNG对象），player（角色实体对象）；返回值类型：无
	TBOMC_PRE_ADD_CARD = "TBOMC_PRE_ADD_CARD",							--回调参数：card（整数）；函数参数：card（整数），player（角色实体对象）；返回值类型：无
	TBOMC_POST_NEW_GREED_MODE_WAVE = "TBOMC_POST_NEW_GREED_MODE_WAVE",	--回调参数：无；函数参数：current_wave（整数）；返回值类型：无
}

tbom.DefaultKeyConfig = {
	["P1 change spell 1 (Keyboard)"] = Keyboard.KEY_LEFT_ALT,
	["P1 change spell 2 (Keyboard)"] = Keyboard.KEY_RIGHT_CONTROL,
	["P1 change spell 1 (Controller)"] = 0,
	["P1 change spell 2 (Controller)"] = 1,

	["P2 change spell 1 (Keyboard)"] = Keyboard.KEY_LEFT_ALT,
	["P2 change spell 2 (Keyboard)"] = Keyboard.KEY_RIGHT_CONTROL,
	["P2 change spell 1 (Controller)"] = 0,
	["P2 change spell 2 (Controller)"] = 1,

	["P3 change spell 1 (Keyboard)"] = Keyboard.KEY_LEFT_ALT,
	["P3 change spell 2 (Keyboard)"] = Keyboard.KEY_RIGHT_CONTROL,
	["P3 change spell 1 (Controller)"] = 0,
	["P3 change spell 2 (Controller)"] = 1,

	["P4 change spell 1 (Keyboard)"] = Keyboard.KEY_LEFT_ALT,
	["P4 change spell 2 (Keyboard)"] = Keyboard.KEY_RIGHT_CONTROL,
	["P4 change spell 1 (Controller)"] = 0,
	["P4 change spell 2 (Controller)"] = 1,
}

tbom.KeyConfig = {}
for key, cfg in pairs(tbom.DefaultKeyConfig) do
	tbom.KeyConfig[key] = tbom.DefaultKeyConfig[key]
end

tbom.TempData = {}
--tbom.TempData.PlayerData_Static = {}
--tbom.TempData.PlayerData_Static["UserNum"] = 0
--tbom.TempData.PlayerData_UserRegister = {}
tbom.TempData.PlayerData = {}
tbom.TempData.GameData = {}
tbom.TempData.NPCData = {}
tbom.TempData.EffectData = {}

tbom.Global = {}
tbom.Global.Common = include("scripts/!global/common")
tbom.Global.Maths = include("scripts/!global/maths")
tbom.Global.Tools = include("scripts/!global/tools")
tbom.Global.Translation = include("scripts/!global/translation")
tbom.Global.ModData = include("scripts/!global/mod_data")

tbom.SpellContent = include("scripts/features/spells/!content")
tbom.Magic = include("scripts/features/magic")

tbom.LevelExp = include("scripts/features/level_exp")

tbom.CriticalChance = include("scripts/features/critical_chance")

local MagicType = tbom.MagicType
tbom.BaseSpell = {
	[MagicType.NONE] = include("scripts/features/spells/000_none/!base"),
	[MagicType.AGGRESSIVE] = include("scripts/features/spells/001_aggressive/!base"),
	[MagicType.DEFENSIVE] = include("scripts/features/spells/002_defensive/!base"),
	[MagicType.SPECIAL] = include("scripts/features/spells/003_special/!base"),
	[MagicType.LOCKON] = include("scripts/features/spells/004_lockon/!base"),
	[MagicType.HELPER] = include("scripts/features/spells/005_helper/!base"),
}

local SpellType = tbom.SpellType
tbom.Spells = {
	--空类型法术
	[SpellType.SPELL_NONE] = include("scripts/features/spells/000_none/000_none"),
	--攻击性法术
	[SpellType.SPELL_FIRE] = include("scripts/features/spells/001_aggressive/001_fire"),
	[SpellType.SPELL_ICE_STORM] = include("scripts/features/spells/001_aggressive/002_ice_storm"),
	[SpellType.SPELL_THUNDER] = include("scripts/features/spells/001_aggressive/003_thunder"),
	--防御性法术
	[SpellType.SPELL_HEALING] = include("scripts/features/spells/002_defensive/004_healing"),
	[SpellType.SPELL_BAYOEN] = include("scripts/features/spells/002_defensive/006_bayoen"),
	[SpellType.SPELL_REVIA] = include("scripts/features/spells/002_defensive/007_revia"),
	--特殊法术
	[SpellType.SPELL_DIACUTE] = include("scripts/features/spells/003_special/005_diacute"),
	--瞄准-锁定型法术
	[SpellType.SPELL_JUGEM] = include("scripts/features/spells/004_lockon/008_jugem"),
}

local modCard = tbom.modCard
tbom.Cards = {
	--命名格式：k[ID]_[名称]
}

local modEffectVariant = tbom.modEffectVariant
tbom.Effects = {
	--命名格式：ef[ID]_[名称]
	[modEffectVariant.EXP] = include("scripts/effects/ef2400_exp/ef2400_exp_main"),
	[modEffectVariant.LASER_SIGHT] = include("scripts/effects/ef2401_laser_sight/ef2401_laser_sight_main"),
	[modEffectVariant.LOCKON_MARK] = include("scripts/effects/ef2402_lockon_mark/ef2402_lockon_mark_main"),
	[modEffectVariant.MAGIC_CIRCLE] = include("scripts/effects/ef2403_magic_circle/ef2403_magic_circle_main"),
	[modEffectVariant.ET_SPACESHIP] = include("scripts/effects/ef2408_et_spaceship/ef2408_et_spaceship_main"),
}

local modFamiliarVariant = tbom.modFamiliarVariant
tbom.Familiar = {
	--命名格式：f[ID]_[名称]
	[modFamiliarVariant.LIGHT_ORB] = include("scripts/familiar/f2400_light_orb/f2400_light_orb_main"),
}

local modEntityType = tbom.modEntityType
tbom.Monsters = {
	--命名格式：e[ID]_[名称]
	[modEntityType.ENTITY_PUYO] = include("scripts/monsters/e305_puyo/e305_puyo_main"),
}

local modCollectibleType = tbom.modCollectibleType
tbom.Collectibles = {
	--命名格式：c[ID]_[名称]
	[modCollectibleType.COLLECTIBLE_BLUE_GRIMOIRE] = include("scripts/items/collectibles/c001_blue_grimoire/c001_blue_grimoire_main"),
	[modCollectibleType.COLLECTIBLE_PUYO_HAT] = include("scripts/items/collectibles/c002_puyo_hat/c002_puyo_hat_main"),
	[modCollectibleType.COLLECTIBLE_WORMHOLE_APPLE] = include("scripts/items/collectibles/c003_wormhole_apple/c003_wormhole_apple_main"),
	[modCollectibleType.COLLECTIBLE_HEART_SHAPED_COOKIE] = include("scripts/items/collectibles/c004_heart_shaped_cookie/c004_heart_shaped_cookie_main"),

	[modCollectibleType.COLLECTIBLE_GREEN_GRIMOIRE] = include("scripts/items/collectibles/c098_green_grimoire/c098_green_grimoire_main"),
}

local modPickupVariant = tbom.modPickupVariant
tbom.Pickups = {
	--命名格式：pick[ID]_[名称]
	[modPickupVariant.PICKUP_MANA] = include("scripts/items/pick ups/pick2400_mana/pick2400_mana_main"),
	[modPickupVariant.PICKUP_GEL] = include("scripts/items/pick ups/pick2401_gel/pick2401_gel_main"),
}

local modTrinketType = tbom.modTrinketType
tbom.Trinkets = {
	--命名格式：t[ID]_[名称]
	[modTrinketType.TRINKET_ANOTHER_CARTRIDE] = include("scripts/items/trinkets/t001_another_cartride/t001_another_cartride_main"),
}

local modPlayerType = tbom.modPlayerType
tbom.Characters = {
	--命名格式：p[ID]_[名称]
	[modPlayerType.PLAYER_ORIGINAL] = include("scripts/characters/p000_original_characters/p000_original_characters_main"),
	[modPlayerType.PLAYER_ARLENADJA] = include("scripts/characters/p001_arle_nadja/p001_arle_nadja_main"),
}

local modChallenge = tbom.modChallenge
tbom.Challenges = {
	--命名格式：ch[ID]_[名称]
	[modChallenge.CHALLENGE_METEOR_SHOWER] = include("scripts/challenges/ch001_meteor_shower/ch001_meteor_shower_main"),
	[modChallenge.CHALLENGE_DESCENDING_INTO_PUYO_HELL] = include("scripts/challenges/ch002_descending_into_puyo_hell/ch002_descending_into_puyo_hell_main"),
}

tbom.CompatibleMods = include("compatible mods/main")

tbom.DebugConsole = include("scripts/debug_console")

TBOM = tbom

function tbom:TryReloadShaders()
	local players = Isaac.FindByType(EntityType.ENTITY_PLAYER)
    if #players <= 0 then
        Isaac.ExecuteCommand("reloadshaders")
    end
end
tbom:AddCallback(ModCallbacks.MC_POST_PLAYER_INIT, tbom.TryReloadShaders)

do
	print("[TBOM] The Binding of Madou v0.1.2 loaded.")
	------开发者专用------
	print("[TBOM] Developer only!!!")
	--"luamod the binding of madou_v0_1_2_dev"
	------
end